File: src/hallsim/view/_bake.py
"""A page baked to a static site: every lever setting on a grid, solved
once and written as JSON beside a page that reads them back.

:func:`bake` is the opt-in alternative to :func:`hallsim.view.serve` for
a page that will live on a static host. Nothing runs after the bake, so
the sliders snap to the grid; the lever rows are what it writes, and the
wiring tab stays with the server.
"""

from __future__ import annotations

import html
import itertools
import json
import logging
import re
import shutil
import time
from pathlib import Path

import numpy as np

from hallsim.view._levers import row_figure, shades_of, spread
from hallsim.view._model import ModelBank, ViewModel, series
from hallsim.view._page import Page
from hallsim.view._theme import (
    CSS,
    CURVE_WIDTH,
    GOOGLE_FONTS,
    HOVER,
    REFERENCE_LINE,
    TOKENS,
    color,
    rgba,
    root_css,
)

log = logging.getLogger(__name__)

#: Panel columns a row is laid out in, wide screens first; the served page
#: picks between the same two by viewport width, and so does the baked one.
COLUMN_BUCKETS = (5, 2)
#: Significant digits a series is written at.
SIGNIFICANT = 5
_HERE = Path(__file__).parent


def clean(value) -> float:
    """A setting value as the page keys it: six decimals, no minus zero."""
    return round(float(value), 6) + 0.0


def setting_key(severities) -> str:
    """The file stem of one setting; ``bake.js`` computes the same."""
    return "_".join(f"{clean(v):.2f}" for v in severities)


def slug(name: str) -> str:
    """A variant name as a directory name."""
    return re.sub(r"[^A-Za-z0-9._-]+", "-", name).strip("-.") or "variant"


def grid(page: Page, step: float) -> list[tuple[float, ...]]:
    """Every setting on the lever grid at ``step``, a preset added when it
    falls between."""
    axes = []
    for lever in page.levers:
        n = int(round((lever.hi - lever.lo) / step))
        axes.append([clean(lever.lo + i * step) for i in range(n + 1)])
    settings = [tuple(s) for s in itertools.product(*axes)]
    seen = set(settings)
    for preset in page.presets.values():
        p = tuple(clean(v) for v in preset)
        if p not in seen:
            settings.append(p)
            seen.add(p)
    return settings


def compact(values) -> list:
    """A series at ``SIGNIFICANT`` digits of its largest value; a value
    that is not finite is written as null."""
    a = np.asarray(values, dtype=float)
    finite = np.isfinite(a)
    top = float(np.max(np.abs(a[finite]))) if finite.any() else 0.0
    if top > 0.0:
        digits = SIGNIFICANT - 1 - int(np.floor(np.log10(top)))
        a = np.round(a, max(digits, 0))
    return [float(v) if ok else None for v, ok in zip(a, finite)]


def _setting_rows(page: Page, vm: ViewModel, ys, sample) -> dict:
    """One setting's panel series: ``series[row][j]`` for a deterministic
    row; ``population[row][j] = {lo, mean, hi}`` across the cells for a
    population row."""
    shown = vm.ts >= 0.0
    out_series: dict[str, list] = {}
    out_pop: dict[str, list] = {}
    for name in page.panels:
        cols = vm.columns[name]
        if name in page.population and sample is not None:
            rows = []
            for c in cols:
                cells = series(sample, shown, c)
                lo, hi = spread(cells)
                rows.append(
                    {
                        "lo": compact(lo),
                        "mean": compact(cells.mean(axis=1)),
                        "hi": compact(hi),
                    }
                )
            out_pop[name] = rows
        else:
            out_series[name] = [compact(series(ys, shown, c)) for c in cols]
    return {"series": out_series, "population": out_pop}


def _layouts(page: Page, variant: str, columns) -> dict:
    """The row figures' layouts per column count, every shaded period in
    under its name so the page keeps the ones a setting switches on."""
    import plotly.io as pio

    zero = tuple(0.0 for _ in page.levers)
    out: dict[str, dict] = {}
    for ncols in columns:
        out[str(ncols)] = {}
        for name in page.panels:
            tall = name in page.population and page.cells
            fig = row_figure(
                page,
                name,
                zero,
                variant,
                height=240 if tall else 210,
                ncols=ncols,
                all_shades=True,
            )
            out[str(ncols)][name] = json.loads(pio.to_json(fig))["layout"]
    return out


def _index(page: Page, step: float, columns) -> dict:
    rows = []
    for name in page.panels:
        col = color(page.colors[name])
        rows.append(
            {
                "name": name,
                "title": page.titles[name],
                "deposit": page.deposits.get(name, ""),
                "color": col,
                "band": rgba(col, 0.22),
                "population": bool(name in page.population and page.cells),
                "panels": [pn.label for pn in page.panels[name]],
            }
        )
    levers = []
    for lever in page.levers:
        handle = page.registry.get(lever.handle)
        levers.append(
            {
                "name": lever.name,
                "handle": lever.handle,
                "lo": lever.lo,
                "hi": lever.hi,
                "signed": lever.lo < 0,
                "description": getattr(handle, "description", "") or "",
            }
        )
    shades = {
        slug(variant): [
            {
                "label": sh.label,
                "negative_label": sh.negative_label,
                "lever": sh.lever,
                "when": sh.when,
                "color": rgba(color(sh.color), sh.opacity * 2.5),
            }
            for sh in shades_of(page, variant)
        ]
        for variant in page.variants
    }
    return {
        "title": page.title,
        "tagline": page.tagline,
        "about": list(page.about),
        "logos": [[f, alt] for f, alt in page.logos],
        "levers": levers,
        "step": step,
        "presets": {
            k: [clean(v) for v in vals] for k, vals in page.presets.items()
        },
        "reference": page.reference,
        "reference_label": page.reference_label,
        "variants": [
            {
                "name": v,
                "slug": slug(v),
                "top": page.variant_label(v)[0],
                "sub": page.variant_label(v)[1],
            }
            for v in page.variants
        ],
        "variant_title": page.variant_title,
        "rows": rows,
        "shades": shades,
        "layouts": {
            slug(v): _layouts(page, v, columns) for v in page.variants
        },
        "style": {
            "reference_line": REFERENCE_LINE,
            "curve_width": CURVE_WIDTH,
            "hover": HOVER,
            "control_band": rgba(TOKENS["color-neutral-700"], 0.18),
        },
        "columns": list(columns),
    }


def _write_page(page: Page, out: Path) -> None:
    import plotly

    plotly_js = Path(plotly.__file__).parent / "package_data" / "plotly.min.js"
    shutil.copy(plotly_js, out / "plotly.min.js")
    shutil.copy(_HERE / "bake.js", out / "bake.js")
    text = (_HERE / "bake.html").read_text()
    text = (
        text.replace("{{TITLE}}", html.escape(page.title))
        .replace("{{FONTS}}", GOOGLE_FONTS)
        .replace("{{CSS}}", root_css() + CSS)
    )
    (out / "index.html").write_text(text)
    if page.logos and page.assets_folder:
        (out / "assets").mkdir(exist_ok=True)
        for f, _ in page.logos:
            src = Path(page.assets_folder) / f
            if src.exists():
                shutil.copy(src, out / "assets" / f)


def _dump(path: Path, payload) -> None:
    path.write_text(json.dumps(payload, separators=(",", ":")))


def bake(
    page: Page,
    out_dir,
    *,
    step: float = 0.25,
    columns=COLUMN_BUCKETS,
    quiet: bool = False,
) -> Path:
    """Write ``page`` as a static site under ``out_dir``: ``index.html``
    with its script and plotly.js; ``index.json`` (levers, presets, rows,
    shaded periods, the row layouts); ``reference/<variant>.json``; and
    ``data/<variant>/<setting>.json`` for every lever setting on the grid
    at ``step``. A population row is sampled at the page's cell count and
    written as the band's edges and mean. Returns the directory."""
    if step <= 0:
        raise ValueError("step must be positive")
    from tqdm import tqdm

    out = Path(out_dir)
    settings = grid(page, step)
    bank = ModelBank(page)
    bank.wait()
    (out / "data").mkdir(parents=True, exist_ok=True)
    (out / "reference").mkdir(exist_ok=True)
    total = len(settings) * len(page.variants)
    log.info(
        "baking %d settings x %d variant(s)", len(settings), len(page.variants)
    )
    started = time.perf_counter()
    done = 0
    bar = tqdm(total=total, disable=quiet, unit="setting")
    for variant in page.variants:
        vm = bank.get(variant)
        folder = out / "data" / slug(variant)
        folder.mkdir(exist_ok=True)
        for sev in settings:
            _, ys = vm.solve(sev)
            sample = vm.population(sev)[0] if vm.n_cells else None
            _dump(
                folder / f"{setting_key(sev)}.json",
                _setting_rows(page, vm, ys, sample),
            )
            vm.discard_population(sev)
            done += 1
            bar.update(1)
            if done == 1:
                per = time.perf_counter() - started
                log.info(
                    "first setting %.1f s; about %.0f min for all %d",
                    per,
                    per * total / 60,
                    total,
                )
        ref = page.presets[page.reference]
        _, ys = vm.solve(ref)
        sample = vm.population(ref)[0] if vm.n_cells else None
        _dump(
            out / "reference" / f"{slug(variant)}.json",
            {
                "t": compact(vm.ts[vm.ts >= 0.0]),
                **_setting_rows(page, vm, ys, sample),
            },
        )
    bar.close()
    _dump(out / "index.json", _index(page, step, columns))
    _write_page(page, out)
    log.info(
        "baked %d settings to %s in %.0f s",
        total,
        out,
        time.perf_counter() - started,
    )
    return out

File: src/hallsim/view/bake.js
/* The baked lever page. index.json describes the page; one JSON per
   variant and slider setting holds the series, fetched as the sliders
   move. The row layouts come baked, every shaded period in under its
   name; the setting decides which stay. */
(async function () {
  "use strict";
  const index = await (await fetch("index.json")).json();
  const app = document.getElementById("app");
  const cache = new Map();
  const refs = new Map();
  const state = {
    variant: index.variants[0].slug,
    sev: index.presets[index.reference].slice(),
    cols: columns(),
  };
  let pending = 0;

  function columns() {
    const wide = index.columns[0];
    const narrow = index.columns[index.columns.length - 1];
    return window.innerWidth > 600 ? wide : narrow;
  }
  function clean(v) {
    return Math.round(Number(v) * 1e6) / 1e6 + 0;
  }
  function key(sev) {
    return sev.map((v) => clean(v).toFixed(2)).join("_");
  }
  function fmt(lever, v) {
    return (lever.signed && v >= 0 ? "+" : "") + Number(v).toFixed(2);
  }
  function el(tag, attrs, children) {
    const node = document.createElement(tag);
    for (const [k, v] of Object.entries(attrs || {})) {
      if (k === "class") node.className = v;
      else if (k === "text") node.textContent = v;
      else if (k.startsWith("on")) node.addEventListener(k.slice(2), v);
      else node.setAttribute(k, v);
    }
    for (const c of children || []) {
      node.appendChild(
        typeof c === "string" ? document.createTextNode(c) : c
      );
    }
    return node;
  }
  function severities() {
    const out = {};
    index.levers.forEach((lv, i) => {
      out[lv.handle] = state.sev[i];
    });
    return out;
  }
  function shadeOn(sh, sev) {
    if (!sh.lever) return true;
    const s = sev[sh.lever] || 0;
    if (sh.when === "positive") return s > 0;
    if (sh.when === "negative") return s < 0;
    return s !== 0;
  }
  function shadeName(sh, sev) {
    const below = (sev[sh.lever] || 0) < 0;
    return sh.negative_label && below ? sh.negative_label : sh.label;
  }

  const brand = el("div", { class: "brand" }, [
    el("h1", { id: "brand", text: index.title }),
  ]);
  if (index.tagline) {
    brand.appendChild(el("div", { class: "tag", text: index.tagline }));
  }
  app.appendChild(el("div", { class: "top" }, [brand]));

  const side = el("div", { class: "side" });
  const chips = el("div", { class: "chips" });
  const chipOf = {};
  for (const name of Object.keys(index.presets)) {
    const chip = el("button", {
      class: "chip",
      text: name,
      onclick: () => setSeverities(index.presets[name]),
    });
    chipOf[name] = chip;
    chips.appendChild(chip);
  }
  side.appendChild(chips);
  const sliders = [];
  const values = [];
  index.levers.forEach((lv, i) => {
    const val = el("span", { class: "val" });
    const title = lv.handle + (lv.description ? ": " + lv.description : "");
    const h3 = el("h3", { title: title }, [lv.name, val]);
    const input = el("input", {
      type: "range",
      min: lv.lo,
      max: lv.hi,
      step: index.step,
      value: state.sev[i],
    });
    input.addEventListener("input", () => {
      val.textContent = fmt(lv, input.value);
    });
    input.addEventListener("change", () => {
      state.sev[i] = clean(input.value);
      render();
    });
    const marks = el("div", { class: "marks" });
    for (let k = 0; k < 5; k++) {
      const m = lv.lo + ((lv.hi - lv.lo) * k) / 4;
      marks.appendChild(el("span", { text: String(Math.round(m * 100) / 100) }));
    }
    side.appendChild(el("div", { class: "card" }, [h3, input, marks]));
    sliders.push(input);
    values.push(val);
  });
  const segOf = {};
  if (index.variants.length > 1) {
    const group = el("div", { class: "seg-group" });
    for (const v of index.variants) {
      const seg = el("button", { class: "seg" }, [
        el("span", { class: "seg-top", text: v.top }),
        el("span", { class: "seg-sub", text: v.sub }),
      ]);
      seg.addEventListener("click", () => {
        state.variant = v.slug;
        render();
      });
      segOf[v.slug] = seg;
      group.appendChild(seg);
    }
    side.appendChild(
      el("div", { class: "card" }, [
        el("h3", { text: index.variant_title }),
        group,
      ])
    );
  }
  if (index.about.length) {
    side.appendChild(
      el("div", { class: "about" }, index.about.map((p) => el("p", { text: p })))
    );
  }
  if (index.logos.length) {
    side.appendChild(
      el(
        "div",
        { class: "logos" },
        index.logos.map(([f, alt]) => el("img", { src: "assets/" + f, alt: alt }))
      )
    );
  }

  const legend = el("div", { class: "legend" });
  const main = el("div", { class: "main" }, [legend]);
  const panelOf = {};
  for (const row of index.rows) {
    const h2 = el("h2", {}, [row.title]);
    if (row.deposit) {
      h2.appendChild(el("span", { class: "dep", text: "  " + row.deposit }));
    }
    const panel = el("div", { class: "panel" });
    panelOf[row.name] = panel;
    main.appendChild(el("div", { class: "row" }, [h2, panel]));
  }
  app.appendChild(el("div", { class: "wrap" }, [side, main]));

  function load(url, store, id) {
    if (!store.has(id)) {
      store.set(
        id,
        fetch(url).then((r) => {
          if (!r.ok) throw new Error(url + " " + r.status);
          return r.json();
        })
      );
    }
    return store.get(id);
  }
  function referenceOf(slug) {
    return load("reference/" + slug + ".json", refs, slug);
  }
  function dataOf(slug, k) {
    return load("data/" + slug + "/" + k + ".json", cache, slug + "/" + k);
  }

  function axes(j) {
    return j === 0 ? {} : { xaxis: "x" + (j + 1), yaxis: "y" + (j + 1) };
  }
  function line(t, y, style, hover, j) {
    const tr = Object.assign(
      { type: "scatter", mode: "lines", x: t, y: y, line: style },
      axes(j)
    );
    if (hover) tr.hovertemplate = hover;
    else tr.hoverinfo = "skip";
    return tr;
  }
  function band(t, lo, hi, fill, j) {
    const edge = (y, tonext) =>
      Object.assign(
        {
          type: "scatter",
          mode: "lines",
          x: t,
          y: y,
          line: { width: 0 },
          fillcolor: fill,
          hoverinfo: "skip",
        },
        tonext ? { fill: "tonexty" } : {},
        axes(j)
      );
    return [edge(hi, false), edge(lo, true)];
  }
  function traces(row, ref, cur) {
    const t = ref.t;
    const s = index.style;
    const current = { color: row.color, width: s.curve_width };
    const out = [];
    row.panels.forEach((_, j) => {
      const pop = cur.population[row.name];
      if (pop) {
        const c = ref.population[row.name][j];
        const p = pop[j];
        out.push(...band(t, c.lo, c.hi, s.control_band, j));
        out.push(...band(t, p.lo, p.hi, row.band, j));
        out.push(line(t, c.mean, s.reference_line, null, j));
        out.push(line(t, p.mean, current, s.hover, j));
      } else {
        const r = ref.series[row.name][j];
        out.push(line(t, r, s.reference_line, null, j));
        out.push(line(t, cur.series[row.name][j], current, s.hover, j));
      }
    });
    return out;
  }
  function layoutFor(row, sev) {
    const base = index.layouts[state.variant][String(state.cols)][row.name];
    const layout = JSON.parse(JSON.stringify(base));
    const on = new Set(
      index.shades[state.variant]
        .filter((sh) => shadeOn(sh, sev))
        .map((sh) => sh.label)
    );
    layout.shapes = (layout.shapes || []).filter((sh) => on.has(sh.name));
    return layout;
  }
  function renderLegend(sev) {
    const items = [
      el("span", {}, [el("span", { class: "ln" }), "current setting"]),
      el("span", {}, [
        el("span", { class: "ln dotted" }),
        index.reference_label,
      ]),
    ];
    for (const sh of index.shades[state.variant]) {
      if (!shadeOn(sh, sev)) continue;
      items.push(
        el("span", {}, [
          el("span", { class: "sw", style: "background:" + sh.color }),
          shadeName(sh, sev),
        ])
      );
    }
    legend.replaceChildren(...items);
  }
  async function render() {
    const k = key(state.sev);
    const slug = state.variant;
    const id = ++pending;
    const sev = severities();
    index.levers.forEach((lv, i) => {
      values[i].textContent = fmt(lv, state.sev[i]);
      sliders[i].value = state.sev[i];
    });
    for (const [name, chip] of Object.entries(chipOf)) {
      const selected = key(index.presets[name]) === k;
      chip.className = selected ? "chip selected" : "chip";
    }
    for (const [s, seg] of Object.entries(segOf)) {
      seg.className = s === slug ? "seg selected" : "seg";
    }
    renderLegend(sev);
    let ref;
    let cur;
    try {
      [ref, cur] = await Promise.all([referenceOf(slug), dataOf(slug, k)]);
    } catch (e) {
      console.error(e);
      return;
    }
    if (id !== pending) return;
    for (const row of index.rows) {
      Plotly.react(
        panelOf[row.name],
        traces(row, ref, cur),
        layoutFor(row, sev),
        { displayModeBar: false, responsive: true }
      );
    }
  }
  function setSeverities(vals) {
    state.sev = vals.map(clean);
    render();
  }
  let resizeTimer = null;
  window.addEventListener("resize", () => {
    clearTimeout(resizeTimer);
    resizeTimer = setTimeout(() => {
      const c = columns();
      if (c !== state.cols) {
        state.cols = c;
        render();
      }
    }, 150);
  });
  render();
})();
